### PR TITLE
Sort strings by locale language

### LINKED ISSUE #14254: Automations sort not taking into account accented characters
### Checklist

- [X] I have updated to the latest available Home Assistant version.
- [X] I have cleared the cache of my browser.
- [X] I have tried a different browser to see if it is related to my browser.

### Describe the issue you are experiencing

When automations have accented characters they are not sorted correctly, this is especially noticeable if the accented character is the first character of the automation name.
This is screenshot is an example where I have an automation name starting with "Água" that appears after those starting with "Teste".
Captura de ecrã 2022-11-01, às 20 37 21
I have tested with other accented characters and the behavior is the same.

### Describe the behavior you expected

The automation named "Água ..." should be the first in the list, in portuguese (and probably in other languages), for sorting purposes "Á" and "A" are equivalent.

### Steps to reproduce the issue

1. Create an automation and name it "Água"
2. Create an automation and name it "Teste"

### What version of Home Assistant Core has the issue?

Home Assistant 2022.10.5

### What was the last working version of Home Assistant Core?

_No response_

### In which browser are you experiencing the issue with?

Google Chrome 105.0.5195.125

### Which operating system are you using to run this browser?

macOS Monterey 12.0.1

### State of relevant entities

_No response_

### Problem-relevant frontend configuration

_No response_

### Javascript errors shown in your browser console/inspector

_No response_

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by kukulich**
## Proposed change

Strings should be sorted by locale language. It will improve sorting mainly for languages with accented letters.

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml

```

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue or discussion:
- Link to documentation pull request:

## Checklist

- [X] The code change is tested and works locally.
- [X] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

I've wanted to write tests but in `formatNumber` test there's this comment: `// Node only ships with English support for Intl, so we can not test for other number formats here.`
So I expect I'm not able to write tests :(

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

File: src/common/string/compare.ts [hunk 1]
```diff
@@ -1,13 +1,8 @@
...
 
-  return 0;
-};
-
-export const caseInsensitiveStringCompare = (a: string, b: string) =>
-  stringCompare(a.toLowerCase(), b.toLowerCase());
+export const caseInsensitiveStringCompare = (
+  a: string,
+  b: string,
+  language: string
```
Comment (bramkragten) [inline src/common/string/compare.ts:7]: Make language optional, when it is not provided, the browser language is used

File: src/data/area_registry.ts [hunk 2]
```diff
@@ -58,7 +58,7 @@ const fetchAreaRegistry = (conn: Connection) =>
     })
     .then((areas) =>
       (areas as AreaRegistryEntry[]).sort((ent1, ent2) =>
-        stringCompare(ent1.name, ent2.name)
+        stringCompare(ent1.name, ent2.name, "en")
```
Comment (bramkragten) [inline src/data/area_registry.ts:61]: Don't use hardcoded languages, if undefined, the browser lang will be used.

File: src/common/string/compare.ts (hunk 1 above)
Comment (kukulich) [reply src/common/string/compare.ts:7]: 👍

**Comment by bramkragten** [comment]
It seems there is no full support for this feature in all browsers: <https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/localeCompare> we need to check for support before using it.

Also for performance, we should consider using https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/Collator

**Comment by kukulich** [comment]
@bramkragten 

> It seems there is no full support for this feature in all browsers

This looks good: https://caniuse.com/localecompare, only old browsers do not have support.

File: src/data/area_registry.ts (hunk 2 above)
Comment (kukulich) [reply src/data/area_registry.ts]: Done.

File: src/common/string/compare.ts (hunk 1 above)
Comment (kukulich) [reply src/common/string/compare.ts]: Done.

File: src/common/string/compare.ts [hunk 3]
```diff
@@ -9,5 +9,26 @@ export const stringCompare = (a: string, b: string) => {
...
 
-export const caseInsensitiveStringCompare = (a: string, b: string) =>
-  stringCompare(a.toLowerCase(), b.toLowerCase());
+export const stringCompare = (
+  a: string,
+  b: string,
+  language: string | undefined = undefined
+) => {
+  try {
+    return new Intl.Collator(language).compare(a, b);
```
Comment (bramkragten) [inline src/common/string/compare.ts:18]: We should not create a new `Collator` for every compare, use a memoize so we cache it

File: src/components/currency-datalist.ts [hunk 4]
```diff
@@ -161,7 +161,7 @@ export const CURRENCIES = [
   "ZWL",
 ];
 
-export const getCurrencyOptions = memoizeOne((language?: string) => {
+export const getCurrencyOptions = memoizeOne((language: string) => {
```
Comment (bramkragten) [inline src/components/currency-datalist.ts:164]:
```suggestion
export const getCurrencyOptions = memoizeOne((language?: string) => {
```

File: src/components/country-datalist.ts [hunk 5]
```diff
@@ -253,7 +253,7 @@ export const COUNTRIES = [
   "ZW",
 ];
 
-export const getCountryOptions = memoizeOne((language?: string) => {
+export const getCountryOptions = memoizeOne((language: string) => {
```
Comment (bramkragten) [inline src/components/country-datalist.ts:256]:
```suggestion
export const getCountryOptions = memoizeOne((language?: string) => {
```

File: src/components/country-datalist.ts [hunk 6]
```diff
@@ -266,14 +266,16 @@ export const getCountryOptions = memoizeOne((language?: string) => {
...
   }));
-  options.sort((a, b) => caseInsensitiveStringCompare(a.label, b.label));
+  options.sort((a, b) =>
+    caseInsensitiveStringCompare(a.label, b.label, language)
+  );
   return options;
 });
 
-export const createCountryListEl = () => {
+export const createCountryListEl = (language: string) => {
```
Comment (bramkragten) [inline src/components/country-datalist.ts:275]:
```suggestion
export const createCountryListEl = (language?: string) => {
```

File: src/data/update.ts [hunk 7]
```diff
@@ -94,23 +97,24 @@ export const filterUpdateEntities = (entities: HassEntities) =>
...
+      language
     );
   });
 
 export const filterUpdateEntitiesWithInstall = (
   entities: HassEntities,
   showSkipped = false
 ) =>
-  filterUpdateEntities(entities).filter((entity) =>
+  filterUpdateEntities(entities, "en").filter((entity) =>
```
Comment (bramkragten) [inline src/data/update.ts:109]:
```suggestion
  filterUpdateEntities(entities).filter((entity) =>
```

File: src/data/update.ts [hunk 8]
```diff
@@ -68,7 +68,10 @@ export const updateReleaseNotes = (hass: HomeAssistant, entityId: string) =>
     entity_id: entityId,
   });
 
-export const filterUpdateEntities = (entities: HassEntities) =>
+export const filterUpdateEntities = (
+  entities: HassEntities,
+  language: string
```
Comment (bramkragten) [inline src/data/update.ts:73]:
```suggestion
  language?: string
```

File: src/common/string/compare.ts [hunk 9]
```diff
@@ -9,5 +9,26 @@ export const stringCompare = (a: string, b: string) => {
...
+  } catch (e) {
```
[the remark this answers is not in the document]
Comment (bramkragten) [reply src/common/string/compare.ts]: yes, but we dont use those (for now)

**Review by bramkragten — APPROVED** [review]

**Pull request merged**